salesagility/SuiteCRM
# Fixes #2787 by checking allowed cron users

This PR (finally) fixes #2787.


## Motivation and Context

### Why we should care
I see this in the forums almost every week. When releases come out and people start upgrading, this problem shows several times a week. I would say all the years we've been suffering from this, and the hours of work consumed by admins and support, all add up to a huge and senseless pain.

### Aren't educational efforts enough?
Sure, people should know how to configure things correctly. However, we're not providing instructions for this in the wiki, there are tons of wrong instructions out there on the Internet, and our installer misleads people.
No, educational efforts are not enough - they are _what we have been failing at_ for all these years. No matter how often I explain all this in the forums, it comes up again and again.

### Why it's not that simple
Because configurations can be very different from system to system, and we need to accomodate them all. Because it's hard to know from within cron, which user your other scripts (the web app) run under. So this information needs to be recorded in one place, then read in another.

## Description

### What it does, and how it impacts people
- It does not affect Windows at all. This is a problem of permissions degradation caused by cron misconfiguration, it was diagnosed in Linux, and the fix affects Linux only (in case you're wondering, `exec('whoami');` works on Windows)
- new installations on Linux will get proper advice as to how to setup cron correctly for their specific case.
- On upgrades, it does not affect people with standardly configured Linux systems (web server user = cron user). These should just go on working.
- On upgrades, people with non-standard Linux configurations will have to review their config.php to keep cron working. They will get explanative log entries and receive instructions about this in the Release notes.

### Technical details
- **on install time**, add an entry to an array in `config.php` called `allowed_cron_users` containing the current web server user. 
- **on install time**, provide correct instructions for cron configuration with that specific user's crontab.
- **on upgrade time**, add the same entry to the same array, if it's not there yet (for people coming from versions of SuiteCRM before this mechanism was introduced)
- for these automatic additions of users to the array, **`root` is treated as a special case**, to discourage its use and to force manual intervention from an admin
- **on Schedulers screen**, show specific correct configuration instructions also
- **when `cron.php` runs**, check if the current user is in the array, and exit if not, leaving appropriate log message 

### Security
Any attack that compromises `cron.php` _or any of the Scheduler scripts that it runs_ will be much more severe if cron is running as `root`, something that happens in a lot of installations out there. We're doing people a favor by nudging them towards a safer configuration.

### Educational efforts

As I said, these are not enough, but they _are_ needed.

#### Suggested text snippet to include in Release Notes:

> New mechanism to check which user cron runs under, to avoid permissions degradation issues. IMPORTANT: after upgrade, check your logs and confirm that your Schedulers are running.


#### Suggested addition to the [wiki installation instructions](https://suitecrm.com/wiki/index.php/Installation):

Insert the following section just before the Title "Upgrading SuiteCRM". This will make it the final section of the previous title ("Installing SuiteCRM").

> "Post-Installation steps"
> 
> To complete installation, the Scheduler jobs need to be enabled. This is done by asking your operating system to run a script called cron.php every minute. This, in turn, manages all SuiteCRM jobs that are supposed to run at scheduled times, supporting functionalities like search indexing, workflows, email notifications, database maintenance, etc.
> 
> In Windows this is setup on the system's "Task Scheduler"; in Linux and iOS, it is setup in crontab. During the installation you will receive detailed instructions with the specific commands you need to run to setup these jobs. After installation, you can still access these instructions in Admin area, under option "Schedulers", at the bottom of the screen.
> 
> To make sure your Scheduler jobs don't create permissions problems, the standard practice is to run them under the same user that your Web server runs under. For different configurations, the allowed_cron_users array in config.php must be adjusted.
> 

#### In the forums

Of course some people will get into trouble with this when upgrading and show up in the forums saying "my cron jobs aren't working", and we'll simply answer: complete your upgrade by following instructions, here is the link to configure cron correctly. We have an admin's attention, he scheduled downtime for this, he's probably even testing on a clone of his production system.

Compare this with what we have now: people come up weeks or months after creating the problem, with complains from end-users and managers, with random symptoms and sometimes varied and extensive system rot in production. After several diagnosis attempts, they are told to chmod their permissions, and the cycle of permissions' degradation starts over, to need fixing again later on...

## How To Test This
I tested these situations:
1. Installation (check `config.php` gets a correct `cron_allowed_users` entry).
2. Upgrade (to check the same as above. I had to create a tweaked upgrade package with the new files).
3. Running cron, both with an allowed user and a non-allowed user.
4. You can also try some variations on the user (root or non-root), etc

## Types of changes

- [ X ] Bug fix (non-breaking change which fixes an issue)
- [ X ] New feature (non-breaking change which adds functionality)
- [ X ] Breaking change (fix or feature that would cause existing functionality to change)

### Final checklist


- [ X ] My code follows the code style of this project found [here](https://suitecrm.com/wiki/index.php/Coding_Standards).
- [ X ] **My change requires a change to the documentation (text suggestions provided).**
- [ X ] I have read the [**How to Contribute**](https://suitecrm.com/wiki/index.php
/Contributing_to_SuiteCRM) guidelines.
